2510: clarifying project identifiers in references

toumorokoshi opened the pull request:
As resource references also include project IDs in GCP, it is valuable to clarify how those should be handled.

In an internal conversation asking about project IDs in resource references, this was clarified to be the guidance for GCP.

Adding this detail that was previously implicit.

Raising one consideration:

Noteably, this may result in a violation of intent: since the request is for all resources to store project numbers for anonymization purposes, storing the project IDs may give a third party or internal Google services enough information to identify some relationship between project numbers and IDs.

toumorokoshi committed:
fix(AIP-2510): clarifying ids in resource refs

As resource references also include project IDs in GCP, it is valuable
to clarify how those should be handled.

shwoodard commented on aip/cloud/2510.md:
```diff
@@ -67,6 +67,46 @@ considered PII and user data, but project numbers are not.
 
+### Resource References
+
+Project IDs also appear in [resource names][], which are stored in other
```
I think, "...which are referred to by other...", would be more accurate as APIs are ephemeral and the whether things are *stored* is outside of the concern of an API surface.

toumorokoshi replied:
rephrased, PTAL.

shwoodard commented on aip/cloud/2510.md:
```diff
@@ -67,6 +67,46 @@ considered PII and user data, but project numbers are not.
+  // project-number: 12345
+  name: "project/12345/books/game-of-tones"
+  shelf: "project/123456/shelves/top-row"
+}
```
Do we express proto values in this form elsewhere?

toumorokoshi replied:
All the other examples that I could find are .proto representations of a message or RPC.

This is similar to the textproto representation found in Google's code base, since I couldn't find an example within AIPs.

toumorokoshi replied:
I could possibly use the capital `Book` in the prefix, which would kind of map to the proto representation of a message.

toumorokoshi replied:
resolved given accepting @shwoodard's PR and making a few follow-up adjustments.

shwoodard commented on aip/cloud/2510.md:
```diff
@@ -87,6 +127,7 @@ organizations and folders.
 
 ## Changelog
 
+- **2022-10-19**: Clarified guidance project identifiers in resource references.
```
Add "for" or "around" between "guidance" and "project"

toumorokoshi replied:
ack.

shwoodard commented on aip/cloud/2510.md:
```diff
@@ -67,6 +67,46 @@ considered PII and user data, but project numbers are not.
+
+Similar to how the guidance around project ID / project number is designed
+to ensure that there is a minimal difference between user input and output,
+the reference value is also untranslated to minimize normalized user input.
```
"reference value" seems like a term that might be defined. What does it mean?

toumorokoshi replied:
I think "resource reference value" might be clearer. I'll fix.

GregFurmanek commented on aip/cloud/2510.md:
````diff
@@ -67,6 +67,46 @@ considered PII and user data, but project numbers are not.
+```proto
+book {
+  // Returning the project ID, since that was what was sent.
+  name: "project/foo/books/game-of-tones"
````
So in the essence you want services to return this string unchanged, correct?

toumorokoshi replied:
this particular line refers to the resource name, which should match the request (project-id or project-number).

The new section does state that resource name in other contexts (e.g. as a resource name) should be unchanged.

GregFurmanek commented on aip/cloud/2510.md:
```diff
@@ -67,6 +67,46 @@ considered PII and user data, but project numbers are not.
 
+### Resource References
+
+Project IDs also appear in [resource names][], which are used
```
Would something like this be simpler?

Project IDs may be represented as:

* project-id: "foo"
* project-number: 12345

Services that return resources with project IDs in their names **should** preserve the Project ID format in the returned response. 

Example:
a client request `projects/foo/books/game-of-tones` would result in:

```proto
book {
  name: "project/foo/books/game-of-tones" // project id is preserved
  shelf: "project/123456/shelves/top-row" // project number can be used
}
```
Note: The project ID preservation is not required for fields other than the resource name.

toumorokoshi replied:
The formatting is nice, but I think the intent is different.

we should be strict with respect to whether the user input (e.g. a project ID or project number based string) is unmodified or not. 

for resource references, they shouldn't be: they should be whatever the user sent.

toumorokoshi replied:
resolving given outdated PR.

shwoodard commented on aip/cloud/2510.md:
```diff
@@ -67,6 +67,46 @@ considered PII and user data, but project numbers are not.
+resource references are also untranslated to minimize normalized user input.
+
+[resource names]: ./0122.md
+[example]: ./0122.md#fields-representing-another-resource
```
Please see how other AIP links are constructed throughout the repo.

toumorokoshi replied:
outdated and fixed!

shwoodard commented on aip/cloud/2510.md:
```diff
@@ -67,6 +67,46 @@ considered PII and user data, but project numbers are not.
+### Resource References
+
+Project IDs also appear in [resource names][], which are used
+when referring to a separate resource ([example][]).
```
"...when referring to another resource..."

shwoodard committed:
feat: edit amendments to 2501

shwoodard committed:
Fix guidance, not about parents

toumorokoshi committed:
fixing a couple examples

- shelf is a separate resource from books entirely (no parent-child
  relationship)
- removing library in the resource name as the hiearchy of book is not
  relevant to illustrate the point of returning back the form sent.

shwoodard commented on aip/cloud/2510.md:
```diff
@@ -67,6 +67,84 @@ considered PII and user data, but project numbers are not.
+include the identifier as it occurred in the request: if the project ID
+was provided it should be returned, and if the project number was
+provided, that is what should be in the response.
+
```
Remove. There are two lines of whitespace. Only one is needed.

shwoodard commented on aip/cloud/2510.md:
````diff
@@ -67,6 +67,84 @@ considered PII and user data, but project numbers are not.
+
+```
+
+if the value of `name` on such a request is,
````
"and if..."

otherwise it doesn't go with sentence that it continues.

shwoodard commented on aip/cloud/2510.md:
````diff
@@ -67,6 +67,84 @@ considered PII and user data, but project numbers are not.
+```
+
+then the value of the field, `name`, returned for the `Book`, should
+match: the project ID and not the number. But, the value for the field,
````
"...not the number should be returned. ..."

Phrases after a colon or semicolon must be complete sentences in their own right, unless they represent a list.

shwoodard commented on aip/cloud/2510.md:
````diff
@@ -67,6 +67,84 @@ considered PII and user data, but project numbers are not.
+shelf: projects/12345/shelves/top-shelf
+```
+
+This guidance is to ensure that there is a minimal difference between user
````
"client" instead of "user" here would be clearer.

toumorokoshi replied:
Every other section refers to the "user" and what the user sent. I actually think "user" is clearer here for that purpose.

What's the rationale for "client" being clearer?

shwoodard replied:
Agreed, consistency is best. Thanks, @toumorokoshi!

GregFurmanek committed:
Merge branch 'master' into toum/resource-ref-project-ids
